crates/fluent-check/src/main.rs

```diff
@@ -0,0 +1,213 @@
+    // These must have translations for every message.
+    let required_langs = [fish_fluent::FALLBACK_LANGUAGE];
+    let fluent_resources = parse_ftl_files(&ftl_dir);
+    check_for_extra_ids(&fluent_resources, &unique_ids);
+    check_for_unsorted_ids(&fluent_resources, &unique_ids);
+    let fluent_bundles = resources_to_bundles(fluent_resources);
+    check_required_langs(&fluent_bundles, &required_langs, &unique_ids);
+}
+
+fn resources_to_bundles(
+    resources: HashMap<String, FluentResource>,
+) -> HashMap<String, FluentBundle<FluentResource>> {
+    let mut bundles = HashMap::new();
+    for (language, resource) in resources {
+        let langid: LanguageIdentifier = language
+            .parse()
+            .map_err(|e| format!("Failed to parse language identifier {language}: {e}"))
+            .unwrap();
+        let mut bundle = FluentBundle::new(vec![langid]);
+        bundle
+            .add_resource(resource)
+            .map_err(|e| {
+                format!("Failed to add FTL resources to the bundle for language {language}: {e:?}")
+            })
+            .unwrap();
+        bundles.insert(language, bundle);
+    }
+    bundles
+}
+
+fn extract_fluent_ids() -> HashSet<String> {
+    let mut id_file_content = String::new();
+    match std::env::var_os("FISH_FLUENT_ID_FILE") {
+        Some(file_path) => {
+            File::open(&file_path)
+                .unwrap_or_else(|_| {
+                    panic!("Failed to open file {file_path:?} (passed via FISH_FLUENT_ID_FILE)")
+                })
+                .read_to_string(&mut id_file_content)
+                .unwrap();
+        }
+        None => {
+            let mut tempfile = fish_tempfile::new_file().unwrap();
+            Command::new(env!("CARGO"))
+                .args([
+                    "check",
+                    "--features=fluent-extract",
+                    "--workspace",
+                    "--all-targets",
+                ])
+                .env("FISH_FLUENT_ID_FILE", tempfile.path().as_os_str())
+                .status()
+                .map_err(|e| format!("Failed to extract Fluent IDs: {e}"))
+                .unwrap();
+            tempfile
+                .get_mut()
+                .read_to_string(&mut id_file_content)
+                .unwrap();
+        }
+    };
+    HashSet::from_iter(id_file_content.lines().map(|line| line.to_string()))
+}
+
+fn parse_ftl_files(ftl_dir: &Path) -> HashMap<String, FluentResource> {
+    let mut bundles = HashMap::new();
+    for dir_entry in ftl_dir.read_dir().unwrap() {
+        let dir_entry = dir_entry.unwrap();
+        let file_name = dir_entry.file_name().into_string().unwrap();
+        let Some(language) = file_name.strip_suffix(".ftl") else {
+            continue;
+        };
+        let mut file_content = String::new();
+        File::open(dir_entry.path())
+            .unwrap()
+            .read_to_string(&mut file_content)
+            .unwrap();
```

Comment:
there's [`std::fs::read_to_string()`](https://doc.rust-lang.org/std/fs/fn.read_to_string.html),
appropriate at least for cases like this where we don't add error context to the `File::open` error.